File: binary_search/sum_of_square_numbers.cs
/*
  Given a non-negative integer c, decide whether there're two integers a and b such that a2 + b2 = c.

  Example:
    Input: c = 5
    Output: true
    Explanation: 1 * 1 + 2 * 2 = 5
*/
public class Solution {
    public bool JudgeSquareSum(int c) {
        long l = 0, r = (long) Math.Sqrt(c);
        while (l <= r) {
            long x = l * l + r * r;
            if (x == c) return true;
            else if (x > c) --r;
            else ++l;
        }

        return false;
    }
}

File: binary_search/sum_of_square_numbers.go
/*
  Given a non-negative integer c, decide whether there're two integers a and b such that a2 + b2 = c.

  Example:
    Input: c = 5
    Output: true
    Explanation: 1 * 1 + 2 * 2 = 5
*/
func judgeSquareSum(c int) bool {
    c64 := int64(c);
    l, r := int64(0), int64(math.Sqrt(float64(c)));
    for l <= r {
        x := l * l + r * r;
        if x == c64 {
            return true;
        } else if x > c64 {
            r--;
        } else {
            l++;
        }
    }

    return false;
}
